core/primitives/src/epoch_manager.rs

```diff
@@ -78,37 +78,77 @@ impl ShardConfig {
+    use_production_config: bool,
+    /// EpochConfig from genesis
     genesis_epoch_config: EpochConfig,
-    simple_nightshade_epoch_config: EpochConfig,
+    /// ShardConfig for the simple nightshade upgrade. Also a temporary implementation that allow us
+    /// to upgrade sharding configuration on mainnet and testnet
+    simple_nightshade_shard_config: Option<ShardConfig>,
 }
 
 impl AllEpochConfig {
     pub fn new(
+        use_production_config: bool,
         genesis_epoch_config: EpochConfig,
         simple_nightshade_shard_config: Option<ShardConfig>,
     ) -> Self {
-        let mut config = genesis_epoch_config.clone();
-        if let Some(ShardConfig {
-            num_block_producer_seats_per_shard,
-            avg_hidden_validator_seats_per_shard,
-            shard_layout,
-        }) = simple_nightshade_shard_config
-        {
-            config.num_block_producer_seats_per_shard = num_block_producer_seats_per_shard;
-            config.avg_hidden_validator_seats_per_shard = avg_hidden_validator_seats_per_shard;
-            config.shard_layout = shard_layout;
-        }
-        Self { genesis_epoch_config, simple_nightshade_epoch_config: config }
+        Self { use_production_config, genesis_epoch_config, simple_nightshade_shard_config }
     }
 
-    pub fn for_protocol_version(&self, protocol_version: ProtocolVersion) -> &EpochConfig {
+    pub fn for_protocol_version(&self, protocol_version: ProtocolVersion) -> EpochConfig {
+        // if SimpleNightshade is enabled, we override genesis shard config with
+        // the simple nightshade shard config
+        let mut config = self.genesis_epoch_config.clone();
         if checked_feature!("stable", SimpleNightshade, protocol_version) {
-            &self.simple_nightshade_epoch_config
-        } else {
-            &self.genesis_epoch_config
-        }
+            if let Some(ShardConfig {
+                num_block_producer_seats_per_shard,
+                avg_hidden_validator_seats_per_shard,
+                shard_layout,
+            }) = &self.simple_nightshade_shard_config
+            {
+                config.num_block_producer_seats_per_shard =
+                    num_block_producer_seats_per_shard.clone();
+                config.avg_hidden_validator_seats_per_shard =
+                    avg_hidden_validator_seats_per_shard.clone();
+                config.shard_layout = shard_layout.clone();
+            }
+        }
+        if self.use_production_config {
+            #[cfg(feature = "protocol_feature_chunk_only_producers")]
+            if checked_feature!(
+                "protocol_feature_chunk_only_producers",
+                ChunkOnlyProducers,
+                protocol_version
+            ) {
+                // On testnet, genesis config set num_block_producer_seats to 200
+                // This is to bring it back to 100 to be the same as on mainnet
+                config.num_block_producer_seats = 100;
```

Comment:
Hm, so one interesting effect here is that *genesis* `num_block_producer_seats` also affects the number of parts we split the chunk into for reed solomon: 

https://github.com/near/nearcore/blob/4ef0bb8b5cbabd265cb15b3ef97ea459f13982a1/nearcore/src/runtime/mod.rs#L868

So today testnet and mainnet use different number of parts. Do we want to fix that as well @mzhangmzz ?